### PR TITLE
Kill feed

### PR DESCRIPTION
**Pull request by TimTheToad**
## Purpose
  - Added killfeed. Kind of self-explaining.
  - SideNote: I intened to have the killfeed textblock animate. With their opacity decreasing over time until removed. However this task was too time-consuming for now so I commented the code that implement this. If this is an issue I can remove it but as a goldplating feature I could give it another try next sprint.

## Changes
 - Added KillFeedGUI UC Class. also made some changes to the HUD.xaml

## Testing
How have one tested the feature to ensure it works?
- [ ] Tested in Sandbox
- [ ] Tested in Multiplayer with 2 clients
- [x] Tested in Multiplayer with more than 2 clients
- [ ] Tested in Benchmark

### CODE REVIEW

File: Assets/NoesisGUI/Xaml/HUD.xaml [hunk 1]
```diff
@@ -79,6 +79,19 @@
 
                 <uc:PromptGUI x:Name="PROMPT" Grid.Row="4" Grid.Column="4" Grid.ColumnSpan="2" HorizontalAlignment="Center" VerticalAlignment="Center"/>
 
+                
+                <!--<Rectangle Grid.Column="4" Grid.ColumnSpan="3" Grid.Row="5" Height="113" Width="335" Fill="#D8FF8B00" Margin="0,0,12,0">
+                    <Rectangle.OpacityMask>
+                        <ImageBrush ImageSource="bar.png"/>
+                    </Rectangle.OpacityMask>
+                </Rectangle> -->
```
Comment (frozenbanana) [inline Assets/NoesisGUI/Xaml/HUD.xaml:87]: unnecessary block of comment?

File: Assets/NoesisGUI/Xaml/HUD.xaml [hunk 2]
```diff
@@ -226,6 +239,12 @@
                 <Image Source="PaintBar.png" Height="150" Width="350" Grid.Column="1" Grid.ColumnSpan="3" Grid.Row="3"/>
                 <Image Source="PaintDrop.png" Height="150" Width="350" Grid.Column="1" Grid.ColumnSpan="3" Grid.Row="3"/>
 
+                <!--<Rectangle Grid.Column="1" Grid.ColumnSpan="3" Grid.Row="3" Height="113" Width="335" Fill="#D8FF8B00" Margin="0,0,12,0">
+                    <Rectangle.OpacityMask>
+                        <ImageBrush ImageSource="PaintDropWhite.png"/>
+                    </Rectangle.OpacityMask>
+                </Rectangle>-->
```
Comment (frozenbanana) [inline Assets/NoesisGUI/Xaml/HUD.xaml:246]: same as above

File: CrazyCanvas/Include/GUI/KillFeedGUI.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	void UpdateFeedTimer(LambdaEngine::Timestamp delta);
+
+	void AddToKillFeed(const LambdaEngine::String& feedMessage, uint8 killedPlayerTeamIndex);
+
+private:
+	void RemoveFromKillFeed(Noesis::TextBlock* textblock);
+
+private:
+	//Noesis::Storyboard* m_pKillFeedStoryBoard			= nullptr;
+	//Noesis::DoubleAnimation* m_pKillFeedDoubleAnimation	= nullptr;
```
Comment (frozenbanana) [inline CrazyCanvas/Include/GUI/KillFeedGUI.h]: same as above

File: CrazyCanvas/Source/GUI/HUDGUI.cpp [hunk 4]
```diff
@@ -389,6 +393,18 @@ void HUDGUI::UpdatePlayerAliveStatus(uint64 UID, bool isAlive)
 	}
 }
 
+void HUDGUI::UpdateKillFeed(const LambdaEngine::String& killed, const LambdaEngine::String& killer, uint8 killedPlayerTeamIndex)
```
Comment (frozenbanana) [inline CrazyCanvas/Source/GUI/HUDGUI.cpp]: `void HUDGUI::UpdateKillFeed(const LambdaEngine::String& killed, const LambdaEngine::String& killer, const uint8 killedPlayerTeamIndex)`
```suggestion
void HUDGUI::UpdateKillFeed(const LambdaEngine::String& killed, const LambdaEngine::String& killer, const uint8 killedPlayerTeamIndex)
```

File: CrazyCanvas/Include/GUI/KillFeedGUI.h [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+	KillFeedGUI();
+	~KillFeedGUI();
+
+	void InitGUI();
+
+	bool ConnectEvent(Noesis::BaseComponent* pSource, const char* pEvent, const char* pHandler) override;
+
+	void UpdateFeedTimer(LambdaEngine::Timestamp delta);
+
+	void AddToKillFeed(const LambdaEngine::String& feedMessage, uint8 killedPlayerTeamIndex);
```
Comment (frozenbanana) [inline CrazyCanvas/Include/GUI/KillFeedGUI.h]:
```suggestion
	void AddToKillFeed(const LambdaEngine::String& feedMessage, const uint8 killedPlayerTeamIndex);
```

File: CrazyCanvas/Source/GUI/KillFeedGUI.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+KillFeedGUI::~KillFeedGUI()
+{
+}
+
+void KillFeedGUI::InitGUI()
+{
+	//m_pKillFeedStoryBoard = FindResource<Noesis::Storyboard>("KillFeedStoryBoard");
+	//m_pKillFeedDoubleAnimation = FindName<Noesis::DoubleAnimation>("KillFeedAnimation");
```
Comment (frozenbanana) [inline CrazyCanvas/Source/GUI/KillFeedGUI.cpp]: If you want to work on the transparency part later then I guess you can keep the comments.

File: CrazyCanvas/Source/GUI/KillFeedGUI.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+			if (pair.second <= 0.0f)
+			{
+				RemoveFromKillFeed(pair.first);
+				m_KillFeedTimers.Erase(m_KillFeedTimers.Begin() + i);
+			}
+		}
+	}
+}
+
+void KillFeedGUI::AddToKillFeed(const LambdaEngine::String& feedMessage, uint8 killedPlayerTeamIndex)
```
Comment (frozenbanana) [inline CrazyCanvas/Source/GUI/KillFeedGUI.cpp]:
```suggestion
void KillFeedGUI::AddToKillFeed(const LambdaEngine::String& feedMessage, const uint8 killedPlayerTeamIndex)
```

File: CrazyCanvas/Source/GUI/KillFeedGUI.cpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+{
+	//LambdaEngine::String name = "FEED" + std::to_string(m_FeedIndex++);
```
Comment (frozenbanana) [inline CrazyCanvas/Source/GUI/KillFeedGUI.cpp]: comment not needed?

File: CrazyCanvas/Source/GUI/KillFeedGUI.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+	//FrameworkElement::GetView()->GetContent()->RegisterName(name.c_str(), feed);
+
+	TextFeedTimer pair = std::make_pair(feed, 8.0f);
+
+	m_KillFeedTimers.EmplaceBack(pair);
+	m_pKillFeedStackPanel->GetChildren()->Add(feed);
+
+	/*m_pKillFeedStoryBoard->SetTargetName(m_pKillFeedDoubleAnimation, name.c_str());
+
+	m_pKillFeedStoryBoard->Begin();*/
```
Comment (frozenbanana) [inline CrazyCanvas/Source/GUI/KillFeedGUI.cpp]: same

**Review by frozenbanana — CHANGES_REQUESTED** [review]
Left some questions and suggestions. Overall it looks good.

File: Assets/NoesisGUI/Xaml/HUD.xaml (hunk 1 above)
Comment (TimTheToad) [reply Assets/NoesisGUI/Xaml/HUD.xaml:87]: Both yes and no, it depends if we will fix the Image opacity bug.

File: Assets/NoesisGUI/Xaml/HUD.xaml (hunk 2 above)
Comment (TimTheToad) [reply Assets/NoesisGUI/Xaml/HUD.xaml:246]: Same as above ;)

File: CrazyCanvas/Include/GUI/KillFeedGUI.h (hunk 3 above)
Comment (TimTheToad) [reply CrazyCanvas/Include/GUI/KillFeedGUI.h]: Yes, should be removed =)

File: CrazyCanvas/Source/GUI/KillFeedGUI.cpp (hunk 6 above)
Comment (TimTheToad) [reply CrazyCanvas/Source/GUI/KillFeedGUI.cpp]: Yeah no, what we have now is fully functional and enough for now. =) I'll remove these

**Review by frozenbanana — APPROVED** [review]
dutti

**Pull request merged**